### ISSUE TITLE
Captions & crossrefs not rendering in Quarto with Flextable

### ISSUE THREAD

**Issue by leeroyaus** [quarto-dev/quarto-cli#1556]
#### Description

When rendered via Quarto with output format as docx, caption (tbl-cap) and crossref (@tbl-ref) not working with flextable.

#### Reproducible example

````
---
title: Flextable in Word
format: docx
---

# Tables

Reference @tbl-flextable

```{r}
#| label: tbl-flextable
#| tbl-cap: "Caption 2"
library(flextable)
flextable(head(mtcars))
```
````

See output docx file: [mwe_book.docx](https://github.com/quarto-dev/quarto-cli/files/9205320/mwe_book.docx)

This issue is also mentioned here on flextable github page: <a href="https://github.com/davidgohel/flextable/issues/385#issuecomment-1196280019">flextable Github page</a>

#### Expected result

Caption and crossref should render correctly in Word doc

#### RStudio information

```bash
$long_version
[1] "2022.02.3+492"

$release_name
[1] "Prairie Trillium"
```

#### OS information
Windows 10 Enterprise

### `quarto check` Output

```bash

[>] Checking Quarto installation......OK
      Version: 0.9.649
      Path: C:\Program Files\Quarto\bin\
      CodePage: 1252

[>] Checking basic markdown render....OK

[>] Checking Python 3 installation....(None)

      Unable to locate an installed version of Python 3.
      Install Python 3 from https://www.python.org/downloads/

[>] Checking R installation...........OK
      Version: 4.2.0
      Path: C:/PROGRA~1/R/R-42~1.0
      LibPaths:
        - C:/Users/username/Documents/R/win-library/4.2
        - C:/Program Files/R/R-4.2.0/library
      rmarkdown: 2.14

[>] Checking Knitr engine render......OK
```

### `quarto tools check` Output

```bash
[>] Inspecting tools

Tool         Status                    Installed     Latest  
chromium     Not installed             ---           869685  
tinytex      External Installation     ---           v2022.07
```

### Checklist

- [X] [formatted your issue](https://yihui.org/issue/#please-format-your-issue-correctly) so it is easier for us to read?
- [X] included a minimal, fully reproducible example in a single .qmd file? Please provide the whole file rather than the snippet you believe is causing the issue.
- [X] documented the quarto version you're running, by pasting the output from running `quarto check` in the "Quarto Check Output" text area?
- [X] documented the version of the quarto tools you're running, by providing the output from running `quarto tools check` in the "Quarto Tools Check Output" text area?
- [X] documented the RStudio IDE version you're running (if applicable), by providing the value displayed in the "About RStudio" main menu dialog?
- [X] documented which operating system you're running? If on Linux, please provide the specific distribution as well.
- [X] upgraded to the latest version, including your versions of R, the RStudio IDE, and relevant R packages?

**Comment by eeholmes** [comment]
I see the same problem with flextable for PDF output also.  

Example on RStudio Cloud https://rstudio.cloud/content/4327822

A similar problem appears with {gt} in PDF. But in this case the caption appears but the Table 1 part does not. It looks ok in Word however.

<img width="370" alt="image" src="https://user-images.githubusercontent.com/2545978/183237031-2d93b06c-8f8e-4ced-a636-8de818721240.png">

<pre>
---
title: "test"
format: 
  pdf: default
  html: default
  docx: default
---

# flextable

Reference @tbl-flextable

```{r}
#| label: tbl-flextable
#| tbl-cap: "Caption 1"
library(flextable)
flextable(head(mtcars))
```

# gt

Reference @tbl-gt

```{r}
#| label: tbl-gt
#| tbl-cap: "Caption 2"
library(gt)
gt(head(mtcars[,1:3]))
```
</pre>

RStudio version on cloud is 2022.02.2 but I see the same problem on my laptop (M1 Mac) with RStudio 2022.07.1+554

**Comment by jjallaire** [comment]
The problem is that flextable is creating raw openxml output for the table which Quarto doesn't know how to parse/interact with (we can do this for raw HTML and raw LaTeX just not raw openxml. 

A possible solution would be to fallback to a div which contains the table and just implements the caption manually (i.e. not actually the openxml table caption but just text appearing above or below). We'll take a look at doing this for our v1.2 release.

**Comment by eeholmes** [comment]
Ah right. That's why flextable works so nicely with Word; it creates raw openxml.

**Comment by bd-karen-byron** [comment]
I'm hoping any caption/crossref solution includes making the auto-numbering work as well  
thanks

**Comment by davidgohel** [comment]
Most of the issues should be now fixed for flextable and quarto (in the github version). The limit for now is in the docx version. There is no caption generated by Quarto as explained upper. flextable manages it but as a result, cross-referencing is not possible with the syntax (@tbl-dudu). If there is a solution later, I will adapt the code. As a workaround, use `officer::run_reference()` and make sure officedown is loaded (to automatically trigger OOXML printing).


``````

---
title: "Untitled"
format: docx
---

```{r include=FALSE}
library(officer)
library(officedown)
knitr::opts_chunk$set(echo = FALSE)
```


blah blah blah `r run_reference(id = "tbl-flextable")`.

```{r}
#| label: tbl-flextable
#| tbl-cap: Caption **strong**
flextable::flextable(mtcars[1:4, 1:4])
```

``````

<img width="1038" alt="Capture d’écran 2022-08-18 à 10 23 15" src="https://user-images.githubusercontent.com/4331618/185347185-ec8ead85-6d18-4e8a-bfe1-c71b161ce23b.png">


The HTML code can contain a caption, but it will be rewritten by Quarto with the values of the R chunk parameters, same for the PDF version. I think that Quarto's support imposes to prefer the values linked to the captions defined in the chunk options rather than the values defined with `set_caption`. I don't have control over the values taken in HTML and PDF, it then seems quite natural to adapt the DOCX behavior (which is not modified by Quarto) to the Quarto behavior, i.e. to prefer the parameters defined in the R chunk. Thanks for telling me if I've misunderstood something, and I'll try to correct my code. With 'R Markdown' documents, it's always the opposite, so I first wanted to implement the same logic for consistency but it proved impossible.

**Comment by jjallaire** [comment]
We will certainly try to maintain strong support for all popular table package. In addition, we are investing aggressively in the [gt](https://gt.rstudio.com/) package (in fact its maintainer @rich-iannone has recently joined the Quarto team to focus on among other things gt and tables more broadly in the Quarto ecosystem).

**Comment by eeholmes** [comment]
Good to know re {gt} package. We (NOAA R users admins) have talked about encouraging a switch to {gt}. But last info I saw (June? 22) https://www.danieldsjoberg.com/gtsummary/articles/rmarkdown.html was that {gt} still not fully supporting docx. Then again, that's from the {gtsummary} vignette not {gt}.

<img width="769" alt="image" src="https://user-images.githubusercontent.com/2545978/185448240-4efe107e-97fa-415e-a252-86dc3d26676b.png">

**Comment by cscheid** [comment]
docx output in gt `main` was merged on July 5th https://github.com/rstudio/gt/pull/962, and will make to the next point release coming up in about 2 weeks.

**Comment by shafayetShafee** [comment]
@cscheid , does that mean if I install the development version right now, I can use `{gt]` for docx output in quarto?

**Comment by cscheid** [comment]
Yes:

````
---
title: docx test
format: docx
---

## docx support in quarto

```{r}
library(gt)
exibble %>% gt()
```
````


<img width="693" alt="image" src="https://user-images.githubusercontent.com/285675/185453878-9e1346e5-45b9-4c83-906e-14ea7de78413.png">

**Comment by cscheid** [comment]
I'm going to push a full fix for this to 1.3, since we'll be adding native AST types for tables, etc. and we should start working on communicating better with downstream libraries.

**Comment by kjell-stattenacity** [comment]
I was able to compile a Word document using Quarto with one flextable.  However, the rendering hangs when I add a second flextable (see example below).  I'm working with the most recent development version of Quarto, and RStudio (2022.07.1) and flextable (flextable_0.8.3.012) are up-to-date. 

---
title: "Untitled"
format: docx
---

{r include=FALSE}
library(officer)
library(officedown)
library(flextable)
knitr::opts_chunk$set(echo = FALSE)

 Table `r run_reference(id = "tbl-flextable")` works fine.

{r}
#| label: tbl-flextable
#| tbl-cap: First Flextable
flextable::flextable(mtcars[1:4, 1:4])

However, the rendering hangs for a second table (Table `r run_reference(id = "tbl-flextable2")`). 

{r}
#| label: tbl-flextable2
#| tbl-cap: Second Flextable
flextable::flextable(mtcars[1:4, 1:4])

**Comment by cscheid** [comment]
@kjell-stattenacity I can reproduce that locally, and that's different, it's actually a bug. Thanks for the report! I'll open a separate issue.

**Comment by davidgohel** [comment]
I think this issue is fixed in flextable since few months. Can this issue be closed?

**Comment by cscheid** [comment]
@davidgohel your previous message said this:

> The limit for now is in the docx version. There is no caption generated by Quarto as explained upper. flextable manages it but as a result, cross-referencing is not possible with the syntax (@tbl-dudu). If there is a solution later, I will adapt the code. As a workaround, use officer::run_reference() and make sure officedown is loaded (to automatically trigger OOXML printing)

I would prefer to keep this issue open so that we can handle the crossref syntax issue better on quarto's end.

**Comment by davidgohel** [comment]
@cscheid OK, thanks. I'd like to have the same behavior as with the other formats. If I should change something with captions in Word for Quarto, let me know (I don't know if I did the right thing!)

**Comment by schwa021** [comment]
> `r run_reference(id = "tbl-flextable")`



> @cscheid , does that mean if I install the development version of `{gt}` right now, I can use `{gt}` for docx output in quarto?

You can produce tables, but, unfortunately, you can't format them. 

[](https://github.com/rstudio/gt/issues/1098)

Hopefully soon?

**Comment by sgvignali** [comment]
Hi all, is there any update for `flextable` and word output?

**Comment by cscheid** [comment]
@sgvignali we're going to do this work in 1.4, as I indicated on the milestone change.

**Comment by cderv** [comment]
For reference to follow work across repo, this is also tracked in **flextable** directly at https://github.com/davidgohel/flextable/issues/494

**Comment by aaronschiff** [comment]
@davidgohel I don't think the workaround you gave above works any more? Here's the Word output I get from the same code with Quarto 1.3.333, Flextable 0.9.1, Officer 0.6.2 and Officedown 0.3.0. It seems to copy the entire table where the cross-reference should be and there are issues with the table captions. 

<img width="627" alt="Screenshot 2023-04-28 at 10 03 55" src="https://user-images.githubusercontent.com/10150126/235001523-36c7e037-9e4a-4795-ab73-78981067c698.png">

**Comment by BMC1986** [comment]
> @davidgohel I don't think the workaround you gave above works any more? Here's the Word output I get from the same code with Quarto 1.3.333, Flextable 0.9.1, Officer 0.6.2 and Officedown 0.3.0. It seems to copy the entire table where the cross-reference should be and there are issues with the table captions.
> 
> <img alt="Screenshot 2023-04-28 at 10 03 55" width="627" src="https://user-images.githubusercontent.com/10150126/235001523-36c7e037-9e4a-4795-ab73-78981067c698.png">

Hi @aaronschiff, did you come up with a work-around for this one? I have the same issue.

**Comment by aaronschiff** [comment]
@BMC1986 no I didn't, sorry. For now I had to stop producing Word output and wait for a future fix in Quarto. PDF output works fine so I am using that in the meantime.

**Comment by ldecicco-USGS** [comment]
Just to clarify, neither flextable nor gt can currently handle cross-referencing in Word, correct? We're waiting on a Quarto update?

That is what I'm noticing but want to make sure I haven't missed a workaround. The only way I can seem to get any table cross-referencing in Word is `knitr::kable` or printing raw markdown with result: as-is.

**Comment by schwa021** [comment]
> Just to clarify, neither flextable nor gt can currently handle cross-referencing in Word, correct? We're waiting on a Quarto update?
> 
> That is what I'm noticing but want to make sure I haven't missed a workaround. The only way I can seem to get any table cross-referencing in Word is `knitr::kable` or printing raw markdown with result: as-is.

Any progress on this? The gt documentation suggests it works - but it definitely doesn't for me.

**Comment by mcanouil** [comment]
@schwa021 since the issue here is really about flex table, please open a new GitHub Discussion with a "working" reproducible example.
(This is to avoid cluttering/diluting this issue).

**Comment by cscheid** [comment]
With the current 1.4 prereleases, the following _almost_ works.

````
---
title: Flextable in Word
format: docx
---

# Tables

Reference @tbl-flextable

```{r}
#| label: tbl-flextable
#| tbl-cap: "Caption 2"
library(flextable)
flextable(head(mtcars))
```
````

Produces a document that Microsoft Word doesn't appear to like. But if you tell it to try to open the file anyway, you should see this:

<img width="541" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/285675/1821913b-d8e5-457f-a0ce-50799d1d7b65">

The remaining issue with Microsoft Word will be handled on #7321.
